### ISSUE TITLE
Implementation of base SGD in vw_sklearn

### ISSUE THREAD

**Issue by RituRajSingh878** [VowpalWabbit/vowpal_wabbit#2339]
I am opening this issue to discuss the implementation of the base SGD in vw_sklearn.
I have opened a pr #2332 for the implementation of the multiclassifier for vw_sklearn and would like to add this base learner also.

These are the links to make a new sklearn API and I will take help from these and also, we have already implemented `BaseEstimator` for classifier and regressor. so that will be also helpful
> http://danielhnyk.cz/creating-your-own-estimator-scikit-learn/
> https://ploomber.io/posts/sklearn-custom/

After the discussion, I will make a pr for the same.

@jackgerrits  @gramhagen

**Comment by gramhagen** [comment]
Is the suggestion to change the base VW class in sklearn_vw or add something new? Can you describe what you're thinking of?

**Comment by RituRajSingh878** [comment]
I was thinking of adding a new class for the SGD. 

https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/linear_model/_stochastic_gradient.py#L68

```julia
class VWSGD(SGDBasEstimator):

    .... Required Code for the wrapper...
```

**Comment by gramhagen** [comment]
I see. I wasn't aware of that class. It does seem like a better base class to use than BaseEstimator. Although it's not part of the public API. Could be because it's newer?

We're trying to support a wide range of sklearn versions so that can make this difficult. Is there functionality that this class provides which we want to add. If so then we can discuss how to do that.

**Comment by RituRajSingh878** [comment]
> I see. I wasn't aware of that class. It does seem like a better base class to use than BaseEstimator. Although it's not part of the public API. Could be because it's newer?
>
Yes, We can use a better class although BaseSGD is not a new class. We can use directly `SGDClassifier` and `SGDRegressor` for the implementation.
https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/linear_model/_stochastic_gradient.py#L1071

> We're trying to support a wide range of sklearn versions so that can make this difficult. Is there functionality that this class provides which we want to add. If so then we can discuss how to do that.

I am thinking that if we are matching VW with sklearn API so if sklearn has different classes for SGD then we can also think of making a new class for SGD. Also, it is the part of Microsoft rl fest [project9](https://vowpalwabbit.org/rlos_projects.html). Also in core `VW` learner, we don't have default use of` SGD` so I think we should make a new class and use the `SGD`  of core VW and match it with sklearn API.

**Comment by gramhagen** [comment]
My understanding is that modules prepended with an underscore are not stable/public. So it would be risky to inherit from that.

Until it becomes stable we may want to just reproduce any missing functionality. I don't see the goal of the sklearn_vw module to duplicate all the sklearn classes, but more as a way to provide a simple wrapper around vw so it can be used like a sklearn model.

That said, if there's something unintuitive or some missing functionality we should focus on that. Otherwise I worry it will be very difficult to support multiple versions of sklearn and keep up with their internal API changes.

**Comment by RituRajSingh878** [comment]
> My understanding is that modules prepended with an underscore are not stable/public. So it would be risky to inherit from that.
> 
agree with this 
> Until it becomes stable we may want to just reproduce any missing functionality. I don't see the goal of the sklearn_vw module to duplicate all the sklearn classes, but more as a way to provide a simple wrapper around vw so it can be used like a sklearn model.
> 
> That said, if there's something unintuitive or some missing functionality we should focus on that. Otherwise I worry it will be very difficult to support multiple versions of sklearn and keep up with their internal API changes.

If we wish to use `sgd` learner for regressor/classifier model then currently we can't use from `sklearn_vw`. So I am suggesting to add this feature so that we can access the `sgd`.

Now we can do this by the addition of code into currently implemented `VW` class in sklearn_vw or we can do it by making a separate class.

**Comment by gramhagen** [comment]
Can you elaborate on what you mean by access the sgd?

**Comment by RituRajSingh878** [comment]
pylibvw.vw wrapper vw
https://github.com/VowpalWabbit/vowpal_wabbit/blob/bcd47ad60dabba63504b6fd8c691b20f1624caf5/python/vowpalwabbit/pyvw.py#L78

sklearn wrapper VW
https://github.com/VowpalWabbit/vowpal_wabbit/blob/bcd47ad60dabba63504b6fd8c691b20f1624caf5/python/vowpalwabbit/sklearn_vw.py#L22
 we can use `sgd` in pylibvw.vw wrapper vw like
```julia
In [39]: model = vw(loss_function='logistic', oaa=3, quiet=True, sgd=True) 
    ...: model.learn('1 | a b c')                                                              

In [40]: model.predict(' | a b c')
```

But we can't use `sgd` in sklearn wrapper VW. So I was talking about the implementation of this.

**Comment by jackgerrits** [comment]
Project 9 in RLOS fest is not tackling usage with SKLearn, but just pyvw proper. Perhaps as a stretch goal SKLearn can be incorporated. As Scott said sklearn_vw's goal in life is to be a wrapper allowing usage of VW in sklearn.

What stops you from adding the sgd parameter to the sklearn vw object and initializing it with that?

**Comment by RituRajSingh878** [comment]
> Project 9 in RLOS fest is not tackling usage with SKLearn, but just pyvw proper. Perhaps as a stretch goal SKLearn can be incorporated. As Scott said sklearn_vw's goal in life is to be a wrapper allowing usage of VW in sklearn.
> 
`Implement a simple gradient descent base learner using SKlearn` this is the part of project 9. So can you please describe it in detail.
> What stops you from adding the sgd parameter to the sklearn vw object and initializing it with that?

https://github.com/VowpalWabbit/vowpal_wabbit/issues/2339#issuecomment-599691731 I have said that We can do this in two ways. I can go with the addition of `sgd` parameter in `VW`.

**Comment by jackgerrits** [comment]
> Implement a simple gradient descent base learner using SKlearn

This means to use SKLearn to implement a single reduction in Python. So instead of using VW's GD reduction it would use a piece of code implementing the model learning and scoring using SKlearn, but adhering the interface of a VW reduction

**Comment by jackgerrits** [comment]
Closed by #2348
